File: read_betsy_2026.py
import numpy as np
import pandas as pd
import read_spo_actris

def read_betsy_2026(filename, STN):
    """
    Equivalent Python version of MATLAB read_betsy_2026
    STN= station name
    """

    # 1. Read data
    station_dat = read_spo_actris(filename)  # assumed DataFrame

    # 2. Convert to numeric (similar to timetable2num)
    station_dat = station_dat.apply(pd.to_numeric, errors='coerce')

    names_var = station_dat.columns.tolist()

    # 3. Station groups
    west = ['BBE','BLI','CRG','GWA','GLR','GBN','GRE','HGC','IBB',
            'MRN','MZW','PAZ','RMN','SCN','SIA','TMO','UBW']
    east = ['ACA','CRO','DSW','GSM','GGW','LBW','MCN','NCC','SHN']

    # helper flags (MATLAB contains bug fixed properly here)
    is_west = STN in west
    is_east = STN in east

    # 4. Scattering thresholds
    if is_west:
        bs_mask = [c for c in names_var if c.startswith(("Bs", "Bbs"))]

        for col in bs_mask:
            station_dat.loc[station_dat[col] > 100, col] = np.nan

        t_mask = [c for c in names_var if c.startswith(("T", "U"))]
        for col in t_mask:
            station_dat.loc[(station_dat[col] >= 9999) |
                (station_dat[col] >= 99999),col] = np.nan

    elif is_east:
        bs_mask = [c for c in names_var if c.startswith(("Bs", "Bbs"))]

        for col in bs_mask:
            station_dat.loc[station_dat[col] > 500, col] = np.nan

        t_mask = [c for c in names_var if c.startswith(("T", "U"))]
        for col in t_mask:
            station_dat.loc[station_dat[col] >= 9999, col] = np.nan

    # 5. Special cases
    if STN == "AMY":
        u_cols = [c for c in names_var if c.startswith("U")]
        for col in u_cols:
            station_dat.loc[station_dat[col] > 990, col] = np.nan

    if STN == "PAL":
        if "BbsG_S11" in station_dat:
            station_dat.loc[station_dat["BbsG_S11"] == 100, "BbsG_S11"] = np.nan
        if "BbsR_S11" in station_dat:
            station_dat.loc[station_dat["BbsR_S11"] == 100, "BbsR_S11"] = np.nan

    # 6. RH filtering and "dry" variables
    u_cols = [c for c in names_var if "U" in c]

    for rh_col in u_cols:

        rh_values = station_dat[rh_col]

        # threshold
        indRH = rh_values > 50

        if STN in ["MCN", "MRN"]:
            indRH = (rh_values > 50) | rh_values.isna()

        # associated scattering variables
        suffix = rh_col[1:]

        assoc_cols = [c for c in names_var if suffix in c and c.startswith("B")]

        for col in assoc_cols:
            dry_col = col + "_dry"
            station_dat[dry_col] = station_dat[col].copy()
            station_dat.loc[indRH, dry_col] = np.nan

    # 7. Daily aggregation
    station_datD = station_dat.resample("D").median()
    station_datN = station_dat.resample("D").count()

    # 8. Data coverage filtering
    bs_cols = [c for c in station_dat.columns if c.startswith("B")]

    a = []

    for col in bs_cols:
        if col in station_datN.columns:

            indN = station_datN[col] < 6

            valid = station_datD[col].notna().sum()
            total_bad = indN.sum()

            if valid > 0:
                a.append(total_bad * 100 / valid)
            else:
                a.append(np.nan)

            station_datD.loc[indN, col] = np.nan

    a = np.array(a)

    if np.any(a > 20):
        print("WARNING: there are many days with low data coverage")

    return station_datD

File: read_betsy_2026.m
function station_datD = read_betsy_2026(filename, STN)
% read the files send by betsy
% filename contains also the whole path
% STN= station name

%exemple:  SPO_rd=read_betsy_2026('/prod/pay/Aerosol_actris_trend/data/spo/spo_1979_2025','SPO');
%% read the data
if strcmp(STN,'SUM2')
    station_dat = importfile_SUM2("C:\github_trend\raw_data\sum_neph_psap_clap2_clap10_AE16_AE33", [2, Inf]);
    station_dat.Properties.DimensionNames{1}='Time';
elseif strcmp(STN,"SPL") || strcmp(STN, "UGR") || strcmp(STN, "APP")   || strcmp(STN, "CPR")
    station_dat = read_SPL_actris(filename, [2, Inf]);
    station_dat.Properties.DimensionNames{1}='Time';
elseif  strcmp(STN, "MLO")
    station_dat = importfile_MLO(filename, [2, Inf]);
elseif  strcmp(STN, "ATTO")
    station_dat = importfile_ATTO(filename, [2, Inf]);
    station_dat.Properties.DimensionNames{1}='Time';
elseif  strcmp(STN, "LLN")
    station_dat = importfile_LLN(filename, [2, Inf]);
    station_dat.Properties.DimensionNames{1}='Time';
elseif  strcmp(STN, "MBO")
    station_dat = importfile_MBO(filename, [2, Inf]);
else
    station_dat= read_spo_actris(filename);
end

%% probleme numeric negative: tentative to solve it
station_dat = timetable2num(station_dat);


%%
names_var=fieldnames(station_dat);
%% remove the variables containing g (STD) and N (nb of data)
c=contains(names_var,{'g','N'});
names_delete=names_var(c);
station_dat=removevars(station_dat,names_delete);
names_var=fieldnames(station_dat);
% Improve station, limit of scattering to 500 Mm-1 for the eastern stations (ACA, GSM, MCN, SHN)
% and 100 Mm-1 for western (BBE HCG IBB MRN MZW PAZ SCN SIA)
%check the RH for the scattering
west=['BBE','BLI','CRG','GWA','GLR','GBN', 'GRE','HGC', 'IBB', 'MRN', 'MZW', 'PAZ', 'RMN','SCN', 'SIA','TMO','UBW'];
east=['ACA', 'CRO','DSW','GSM','GGW','LBW', 'MCN','NCC', 'SHN'];
if contains(west,STN)
    CW=startsWith(names_var,["Bs","Bbs"]);
    N=names_var(CW);
    for i=1:length(N)
        ind=station_dat.(N{i})>100;
        station_dat.(N{i})(ind)=NaN;
    end
    Cc=startsWith(names_var,["T","U"]);
    N=names_var(Cc);
    for i=1:length(N)
        ind=station_dat.(N{i})>=9999 | station_dat.(N{i})>=99999 | station_dat.(N{i})>=99999;
        station_dat.(N{i})(ind)=NaN;
    end
elseif contains(east,STN)
    CE=startsWith(names_var,["Bs","Bbs"]);
    N=names_var(CE);
    for i=1:length(N)
        ind=station_dat.(N{i})>500;
        station_dat.(N{i})(ind)=NaN;
    end
    Cc=startsWith(names_var,["T","U"]);
    N=names_var(Cc);
    for i=1:length(N)
        ind=station_dat.(N{i})>=9999;
        station_dat.(N{i})(ind)=NaN;
    end
elseif strcmp(STN,'SUM2')

    Cc=startsWith(names_var,["B","U","X"]);
    N=names_var(Cc);
    for i=1:length(N)
        ind=station_dat.(N{i})>=9999;
        station_dat.(N{i})(ind)=NaN;
    end
else
    Cc=startsWith(names_var,["B","T","U"]) & ~startsWith(names_var,'Time');
    N=names_var(Cc);
    for i=1:length(N)
        ind=station_dat.(N{i})>=9999 ;
        station_dat.(N{i})(ind)=NaN;
    end
end

%specail reading
%erreur special pour AMY
if strcmp(STN,'AMY')==1
    Cu=startsWith(names_var,'U');
    Nu=names_var(Cu);
    for i=1:length(Nu)
        ind=station_dat.(Nu{i})>990;
        station_dat.(Nu{i})(ind)=NaN;
    end

    %erreur special pour PAL: false missing code for G and R Bbs
elseif strcmp(STN,'PAL')==1
    station_dat.BbsG_S11(station_dat.BbsG_S11==100)=NaN;
    station_dat.BbsR_S11(station_dat.BbsR_S11==100)=NaN;

    % time treatment for SPL
elseif strcmp(STN,'SPL')==1 || strcmp(STN,'UGR')==1  || strcmp(STN,'ATTO')==1  || strcmp(STN,'APP')==1 || strcmp(STN,'CPR')==1 

    station_dat(:,1)=[];
    station_dat.Properties.VariableNames{1}='y';
    station_dat.DOY=[];
    station_dat=table2timetable(station_dat);
    station_dat.Properties.DimensionNames{1}='Time';
    Cu=startsWith(names_var,["U";"B"]);
    Nu=names_var(Cu);
    for i=1:length(Nu)
        ind=station_dat.(Nu{i})>880 | station_dat.(Nu{i})<=-99;
        station_dat.(Nu{i})(ind)=NaN;
    end
elseif strcmp(STN,'LLN')==1

    station_dat(:,1)=[];
    station_dat.Properties.VariableNames{1}='y';
%    station_dat=table2timetable(station_dat);

elseif strcmp(STN, 'MBO')
    station_dat.MBO=[];
    station_dat.Properties.VariableNames{1}='y';
    station_dat.DOY=[];
    station_dat=table2timetable(station_dat);
    station_dat.Properties.DimensionNames{1}='Time';
    Cu=startsWith(names_var,["U";"B"]);
    Nu=names_var(Cu);
    for i=1:length(Nu)
        ind=station_dat.(Nu{i})>880 | station_dat.(Nu{i})<=-99;
        station_dat.(Nu{i})(ind)=NaN;
    end
elseif strcmp(STN,'MLO')==1
    station_dat.Properties.DimensionNames{1}='Time';
end

names_var=fieldnames(station_dat);
CU=contains(names_var,'U');
RH=names_var(CU);
for i=1:length(RH)
    N=RH{i}(2:end);
    %select RH too high
    if strcmp(STN,'GSN')==1 && strcmp(RH{i},'U_S11')
        RH{i}='U0_S11';

    end
    indRH=station_dat.(RH{i})>50 ;
    if strcmp(STN,'MCN')==1 | strcmp(STN,'MRN')==1
        indRH=station_dat.(RH{i})>50  | isnan(station_dat.(RH{i}));
    end
    % select variables associated with RH measurements
    C1=contains(names_var,RH{i}(2:end));
    CB=startsWith(names_var,'B');
    Cok=C1 & CB;
    S1=names_var(Cok);
    %special case of fragmentation of GSN scat but not RH

    for j=1:length(S1)
        S=strcat((S1{j}),'_dry');
        station_dat.(S)=station_dat.(S1{j});
        station_dat.(S)(indRH)=NaN;
    end
end


%test RMN without negatives
% % if strcmp(STN,'RMN')
% %     SR=names_var(C1);
% %     for i=1:length(SR)
% %         indRMN=station_dat.(SR{i})<0;
% %         station_dat.(SR{i})(indRMN)=NaN;
% %     end
% % end
% daily mean, apply a data coverage threshold ?
station_datD=retime(station_dat,'daily',@nanmedian);
station_datN=retime(station_dat,'daily','count');
names_var=fieldnames(station_dat);
CB=startsWith(names_var,'B');
SB=names_var(CB);
for i=1:length(SB)
    indN=station_datN.(SB{i})<6 ;
    a(i)=sum(indN)*100/sum(~isnan(station_datD.(SB{i})));
    station_datD.(SB{i})(indN)=NaN;
end
ind=a>20;
if sum(ind)>0
    warning('there is a lot of days with less than 50% data coverage');

end


end

